cvmfs/export_plugin/fs_traversal_interface.h

```diff
@@ -0,0 +1,346 @@
+
+#include "libcvmfs.h"
+
+#define COPY_BUFFER_SIZE 4096
+
+struct fs_traversal_context {
+  uint64_t version;
+  uint64_t size;
+
+  char *repo;
+  char *base;
+  char *data;
+  char *config;
+  char *lib_version;
+
+  void * ctx;
+};
+
+enum fs_open_type {
+  fs_open_read = 1,
+  fs_open_write = 2,
+  fs_open_append = 4
+};
+
+struct fs_traversal {
+  /**
+   * Local copy of the context that should be set with the
+   * returned value of initialize. Ideally this is created
+   * and then freed all within the traversal.
+   */
+  struct fs_traversal_context *context_;
+
+  /**
+   * Method which initializes a file system traversal context based on
+   * repository and data storage information
+   * 
+   * @param[in] repo Repository name
+   * @param[in] base Base repository location
+   * @param[in] data Data storage information
+   * @param[in] num_threads Number of threads that may be started
+   * @param[in] config A configuration file name
+   * @param[out] A pointer to the context
+   */
+  struct fs_traversal_context *(*initialize)(
+    const char *repo,
+    const char *base,
+    const char *data,
+    int num_threads,
+    const char *config);
+
+  /**
+   * Method which finalizes the file system traversal.
+   * Must always be called after a context had been initialized and before the
+   * procedure ends
+   * 
+   * The context is freed during execution of this method
+   * 
+   * @param[in] ctx The context to finalize and free
+   */
+  void (*finalize)(struct fs_traversal_context *ctx);
+
+  /**
+   * Method that takes the specifications provided and stores
+   * them in a provenance directory. 
+   *
+   * @param[in] src The src context being archived
+   * @param[in] dest The dest context being archived
+   */
+  void (*archive_provenance)(struct fs_traversal_context *src,
+                             struct fs_traversal_context *dest);
+
+  /**
+   * Method which returns a list over the given directory
+   * 
+   * @param[in] ctx The file system traversal context
+   * @param[in] dir The directory over which should be iterated
+   * @param[out] buf The list of the paths to the elements in the directory
+   * @param[out] len Length of the output list
+   */
+  void (*list_dir)(struct fs_traversal_context *ctx,
+                const char *dir,
+                char ***buf,
+                size_t *len);
+
+  /**
+   * Method which returns a stat struct given a file path.
+   * If the file doesn't exist, NULL is returned
+   * 
+   * @note(steuber): The content checksum only needs to be calculated if this
+   * is a source file system!
+   * 
+   * @param[in] ctx The file system traversal context
+   * @param[in] path The path of the object to stat
+   * @param[out] stat The stat struct to write the information into
+   * @param[in] get_hash Whether the file hash
+   *  should be calculated (true) or not (false)
+   * @returns 0 on success
+   */
+  int (*get_stat)(struct fs_traversal_context *ctx,
```

Comment:
Update description to specify values not allowed NULL. Such as cvm_name.

Reply:
added something